biz.aQute.bndlib/src/aQute/bnd/build/Workspace.java

```diff
@@ -1521,16 +1548,29 @@ Repository getResourceRepository() throws Exception {
 
 	/**
 	 * A macro that returns a set of resources in bundle selection format from
-	 * the repository. For example:
+	 * the repositories. For example:
 	 *
 	 * <pre>
 	 * ${findproviders;osgi.wiring.package;(osgi.wiring.package=aQute.bnd.build)}
 	 * </pre>
 	 */
 
 	public String _findproviders(String[] args) throws Exception {
-		Macro.verifyCommand(args, _findprovidersHelp, null, 2, 3);
-		return findProviders(args[1], args.length > 2 ? args[2] : null).map(Capability::getResource)
+		Macro.verifyCommand(args, _findprovidersHelp, null, 2, 4);
+		return findProvidersMacro(args);
+	}
+
+	private enum STRATEGY {
+		ALL,
+		REPOS,
+		WORKSPACE
+	}
+
+	public String findProvidersMacro(String[] args) throws Exception {
+		String namespace = args[1];
+		String filter = args.length > 2 ? args[2] : null;
+		STRATEGY strategy = args.length > 3 ? STRATEGY.valueOf(args[3]) : STRATEGY.ALL;
```

Comment:
Using `valueOf` here is a bit problematic since it does not allow the macro writer to use lower case and on misspellings we get an exception. See https://github.com/bndtools/bnd/blob/df01bd872b34497d9cae0a7a7eb8fd4a3c417dab/biz.aQute.bndlib/src/aQute/bnd/build/Project.java#L1637-L1644

Reply:
Done. I've given the messages Interface a shot as well. I hoe I did it correctly.